app/src/main/java/ai/elimu/appstore/synchronization/DownloadApplicationAsyncTask.java

```diff
@@ -0,0 +1,235 @@
+                    return null;
+                }
+
+//                    byte[] bytes = IOUtils.toByteArray(inputStream);
+                ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
+                byte[] buffer = new byte[1024];
+                int bytesRead = 0;
+                while ((bytesRead = inputStream.read(buffer)) != -1) {
+                    byteArrayOutputStream.write(buffer, 0, bytesRead);
+
+                    fileSizeInKbsDownloaded += (bytesRead / 1024);
+
+                    //Only update APK size every 100KB downloaded
+                    if (fileSizeInKbsDownloaded % 100 == 0) {
+                        publishProgress(fileSizeInKbsDownloaded);
+                    }
+                }
+                byte[] bytes = byteArrayOutputStream.toByteArray();
+
+                fileOutputStream = new FileOutputStream(apkFile);
+                fileOutputStream.write(bytes);
+                fileOutputStream.flush();
+            } catch (MalformedURLException e) {
+                Timber.e(e, "MalformedURLException");
+            } catch (IOException e) {
+                Timber.e(e, "IOException");
+            } finally {
+                if (fileOutputStream != null) {
+                    try {
+                        fileOutputStream.close();
+                    } catch (IOException e) {
+                        Timber.i(e, "IOException");
+                    }
+                }
+            }
+        }
+
+        Timber.i("apkFile: " + apkFile);
+
+        return fileSizeInKbsDownloaded;
+    }
+
+    @Override
+    protected void onProgressUpdate(Integer... fileSizeInKbsDownloadeds) {
+        Timber.d("onProgressUpdate");
+        super.onProgressUpdate(fileSizeInKbsDownloadeds);
+
+        int fileSizeInKbsDownloaded = fileSizeInKbsDownloadeds[0];
+        Timber.d("fileSizeInKbsDownloaded: " + fileSizeInKbsDownloaded);
+
+        int progress = (fileSizeInKbsDownloaded * 100) / applicationVersion.getFileSizeInKb();
+        Timber.d("progress: " + progress);
+        mProgressBarDownloadProgress.setProgress(progress);
+
+        // E.g. "6.00 MB/12.00 MB   50%"
+
+        String progressText = String.format(mContext.getString(R.string
+                        .app_list_download_progress_number), fileSizeInKbsDownloaded / 1024f,
+                applicationVersion.getFileSizeInKb() / 1024f, progress);
+
+        Timber.d("progressText: " + progressText);
+        mTextViewDownloadProgress.setText(progressText);
+    }
+
+    @Override
+    protected void onPostExecute(Integer fileSizeInKbsDownloaded) {
+        Timber.i("onPostExecute");
+        super.onPostExecute(fileSizeInKbsDownloaded);
+
+        // Hide progress indicators
+        mProgressBarDownloadProgress.setVisibility(View.GONE);
+        mTextViewDownloadProgress.setVisibility(View.GONE);
+
+        if (fileSizeInKbsDownloaded == null || fileSizeInKbsDownloaded == 0) {
```

Comment:
To ease readability, use parenthesis around the boolean expressions when there are multiple.